llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+							}
+						}
+						else
+							next = BI.getSuccessor(0);
+						break;
+					}
+					case Instruction::Switch:
+					{
+						const SwitchInst& SI = cast<SwitchInst>(I);
+				
+						if (isValueComputed(SI.getCondition()))
+						{
+							GenericValue V = getOperandValue(SI.getCondition());
+							ConstantInt* CI = (ConstantInt*)ConstantInt::get((IntegerType*)(SI.getCondition()->getType()), V.IntVal);
+							next = SI.findCaseValue(CI)->getCaseSuccessor();
+						}
+						break;
+					}
+					case Instruction::Ret:
+					{
+						const ReturnInst& RI = cast<ReturnInst>(I);
+
+						if (RI.getReturnValue())
+							stronglyKnownBits[getFirstStack().Caller] = getBitMask(RI.getReturnValue());
+
+						//Dispatch to the Interpreter's visitor for the given Instructon
+						Interpreter::visit(const_cast<Instruction&>(I));
+
+						//Note that here we return
+						return;
+					}
+					default:
+					{
+						//Default handling of not doing anything is conservative
+						break;
+					}
+				}
+
+				if (next)
+				{
+					//We know where execution should proceed
+					setIncomingBB(I.getParent());
+					getLastStack().CurBB = const_cast<BasicBlock*>(next);
+					getLastStack().CurInst = getLastStack().CurBB->begin();
+					//Also here we have set the proper state for the execution so we can return
+					return;
+				}
+				else
+				{
+					skip = true;
+				}
+			}
+
+			//We are inside a call, here we assume all failure to execute are non-recoverable (as in no information could be gained)
+			if (skip)
+			{
+				//Pop current stack (will happen recursively over the call-stack)
+				popSingleStack();
+				// TODO(carlo): remove also stronglyKnownBits
+				removeFromMaps(getLastStack().Caller);
+				return;
+			}
+		}
+		//If we are here it means we have to actually perform the execution via Interpreter
+
+		//Dispatch to the Interpreter's visitor for the given Instructon
+		Interpreter::visit(const_cast<Instruction&>(I));
+
+		//Add  knownBits information
+		assignToMaps(&I, computeStronglyKnownBits(I));
+	}
+
+	/// Create a new interpreter object.
+	///
+	static ExecutionEngine* create(std::unique_ptr<Module> M, std::string *ErrStr)
+	{
+		// Tell this Module to materialize everything and release the GVMaterializer.
+		if (Error Err = M->materializeAll()) {
+			std::string Msg;
+			handleAllErrors(std::move(Err), [&](ErrorInfoBase &EIB) {
+					Msg = EIB.message();
+					});
+			if (ErrStr)
+				*ErrStr = Msg;
+			// We got an error, just return 0
+			return nullptr;
+		}
+
+		return new PartialInterpreter(std::move(M));
+	}
+};
+
+static void removeEdgeBetweenBlocks(llvm::BasicBlock* from, llvm::BasicBlock* to)
+{
+	//Given two BasicBlocks, remove all edges between them.
+	//Multiple edges are legal and might happen since SwitchInst might have more that one case going to the same BB
+
+	{
+		//First count how many edges there are betwen from and to
+		uint32_t cardinality = 0;
+		for (BasicBlock* bb : predecessors(to))
```

Comment:
I think iterating on from successors is cleaner. Also, use a lambda to get the cardinality value instead of this free standing blocks.